lib/spack/spack/ci.py

```diff
@@ -852,73 +923,47 @@ def generate_gitlab_ci_yaml(
                     tty.warn("No match found for {0}, skipping it".format(release_spec))
                     continue
 
-                tags = [tag for tag in runner_attribs["tags"]]
-
-                if spack_pipeline_type is not None:
-                    # For spack pipelines "public" and "protected" are reserved tags
-                    tags = _remove_reserved_tags(tags)
-                    if spack_pipeline_type == "spack_protected_branch":
-                        tags.extend(["protected"])
-                    elif spack_pipeline_type == "spack_pull_request":
-                        tags.extend(["public"])
-
-                variables = {}
-                if "variables" in runner_attribs:
-                    variables.update(runner_attribs["variables"])
-
-                image_name = None
-                image_entry = None
-                if "image" in runner_attribs:
-                    build_image = runner_attribs["image"]
-                    try:
-                        image_name = build_image.get("name")
-                        entrypoint = build_image.get("entrypoint")
-                        image_entry = [p for p in entrypoint]
-                    except AttributeError:
-                        image_name = build_image
-
-                job_script = ["spack env activate --without-view ."]
+                job_object = {
+                    "stage": stage_name,
+                    "variables": {
+                        "SPACK_JOB_SPEC_DAG_HASH": release_spec_dag_hash,
+                        "SPACK_JOB_SPEC_PKG_NAME": release_spec.name,
+                    },
+                    "script": [
+                        "spack env activate --without-view .",
+                        "spack ci rebuild",
+                    ],
+                    "tags": [],
+                    "artifacts": {
+                        "paths": [],
+                        "when": "always",
+                    },
+                    "needs": [],
+                    "retry": {
+                        "max": 2,
+                        "when": JOB_RETRY_CONDITIONS,
+                    },
+                    "interruptible": True,
+                }
 
                 if artifacts_root:
-                    job_script.insert(0, "cd {0}".format(concrete_env_dir))
-
-                job_script.extend(["spack ci rebuild"])
-
-                if "script" in runner_attribs:
-                    job_script = [s for s in runner_attribs["script"]]
-
-                before_script = None
-                if "before_script" in runner_attribs:
-                    before_script = [s for s in runner_attribs["before_script"]]
-
-                after_script = None
-                if "after_script" in runner_attribs:
-                    after_script = [s for s in runner_attribs["after_script"]]
+                    job_object["script"].insert(0, "cd {0}".format(concrete_env_dir))
```

Comment:
Maybe I'm reading this wrong?  It looks like it will update the outgoing job object with the command, regardless of whether the user supplied their own script.  While before this, the `job_script` list that was updated here might have been completely overwritten with the users script list elements just below.

Reply:
With this PR the order of operations is reversed, the whole `job_object` is generated first and then the user's modifications are applied afterwards. This ensures the user's configuration takes precedence in all cases and for all fields. It also keeps the modification order consistent and unsurprising (e.g. `+script` appends to the end of the default `script`, as one might expect). This pattern is repeated throughout the PR for the other job types as well.

https://github.com/spack/spack/pull/32300/files#diff-49f4a785e6a8295a5cd5af6a072cde881c00aecefe40fadf40abb33d5c35f6f5R1148-R1150